File: apps/ios/Exerly/Features/Profile/PreferencesStore.swift
import Foundation

struct PreferencesSnapshot: Codable {
    let schemaVersion: Int
    let accountID: String
    let revision: Int
    var values: [String: JSONValue]
    let user: UserDTO

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version", accountID = "account_id"
        case revision, values, user
    }

    func value(_ key: String) -> JSONValue {
        let parts = key.split(separator: ".").map(String.init)
        if parts.count == 2, case .object(let group) = values[parts[0]] {
            return group[parts[1]] ?? .null
        }
        return values[key] ?? .null
    }

    var heightCM: Double? {
        if case .number(let value) = values["height"] { return value }
        return nil
    }
}

enum PreferenceFields {
    static let definitions: [(key: String, label: String)] = [
        ("name", "Name"), ("age", "Age"), ("gender", "Gender identity"),
        ("unitSystem", "Display units"), ("height", "Height"),
        ("activityLevel", "Usual activity level"), ("timezone", "Time zone"),
        ("dietaryStyle", "Diet preference"), ("allergies", "Allergies"), ("mealsPerDay", "Meals per day"),
        ("experienceLevel", "Training experience"), ("equipmentAccess", "Training location"),
        ("equipment", "Available equipment"), ("activityTypes", "Preferred activities"),
        ("workoutDaysPerWeek", "Weekly workout goal"), ("workoutDays", "Workout days"),
        ("sleepGoalHours", "Sleep goal (hours)"), ("bedtime", "Preferred bedtime"), ("wakeTime", "Preferred wake time"),
        ("reminders.meals", "Meal reminders"), ("reminders.workouts", "Workout reminders"), ("reminders.sleep", "Sleep reminders"),
        ("reminderTimes.meals", "Meal reminder times"), ("reminderTimes.workout", "Workout reminder time"),
        ("reminderTimes.sleep", "Sleep reminder time")
    ]
    static let days = ["monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"]
    static let lists = ["allergies", "equipment", "activityTypes", "workoutDays", "reminderTimes.meals"]
    static let numbers = ["age", "mealsPerDay", "workoutDaysPerWeek", "sleepGoalHours"]
    static func label(_ key: String) -> String { definitions.first { $0.key == key }?.label ?? key.capitalized }

    static func numberText(_ number: Double) -> String {
        number.formatted(.number.grouping(.never).precision(.fractionLength(0...8)))
    }
    static func heightText(_ cm: Double?, units: String) -> String {
        guard let cm else { return "" }
        return numberText(units == "imperial" ? cm / 2.54 : cm)
    }
    static func text(_ value: JSONValue) -> String {
        switch value {
        case .null: return ""
        case .string(let value): return value
        case .number(let value): return numberText(value)
        case .bool(let value): return String(value)
        case .array(let values): return values.map(text).joined(separator: "\n")
        case .object(let value): return value.keys.sorted().map { "\($0): \(text(value[$0]!))" }.joined(separator: "; ")
        }
    }
    static func describe(_ value: JSONValue) -> String {
        if case .bool(let enabled) = value { return enabled ? "On" : "Off" }
        let description = text(value).replacingOccurrences(of: "\n", with: ", ")
        return description.isEmpty ? "Not set" : description
    }
    static func fields(_ snapshot: PreferencesSnapshot) -> [String: String] {
        var result = Dictionary(uniqueKeysWithValues: definitions.map { key, _ in
            (key, key.hasPrefix("reminders.") ? String(snapshot.value(key) == .bool(true)) : text(snapshot.value(key)))
        })
        result["height"] = heightText(snapshot.heightCM, units: result["unitSystem"] ?? "")
        return result
    }
    static func set(_ value: JSONValue, key: String, in values: inout [String: JSONValue]) {
        let parts = key.split(separator: ".").map(String.init)
        if parts.count == 2 {
            var group: [String: JSONValue] = [:]
            if case .object(let current) = values[parts[0]] { group = current }
            group[parts[1]] = value
            values[parts[0]] = .object(group)
        } else { values[key] = value }
    }
    static func changes(_ draft: PreferencesDraft) throws -> [String: JSONValue] {
        let before = fields(draft.base)
        var result: [String: JSONValue] = [:]
        for (key, label) in definitions {
            let raw = (draft.fields[key] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            if key == "height" {
                if !raw.isEmpty && UserEnteredNumber.parse(raw) == nil { throw PreferencesError.message("Enter a valid height.") }
                if draft.heightCM != draft.base.heightCM { result[key] = draft.heightCM.map(JSONValue.number) ?? .null }
                continue
            }
            guard draft.fields[key] != before[key] else { continue }
            let value: JSONValue
            if numbers.contains(key) {
                if raw.isEmpty { value = .null } else if let number = UserEnteredNumber.parse(raw) { value = .number(number) } else { throw PreferencesError.message("Enter a valid number for \(label.lowercased()).") }
            } else if lists.contains(key) {
                var seen = Set<String>()
                value = .array(raw.components(separatedBy: .newlines)
                    .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                    .filter { !$0.isEmpty && seen.insert($0).inserted }.map(JSONValue.string))
            } else if key.hasPrefix("reminders.") { value = .bool(raw == "true") } else { value = raw.isEmpty ? .null : .string(raw) }
            set(value, key: key, in: &result)
        }
        return result
    }
}

struct PreferenceMutation: Codable {
    let baseRevision: Int
    let changes: [String: JSONValue]
    enum CodingKeys: String, CodingKey { case baseRevision = "base_revision", changes }
}
struct PendingPreferences: Codable {
    let id: String
    let body: PreferenceMutation
}
struct PreferencesDraft: Codable {
    let version: Int
    let accountID: String
    var base: PreferencesSnapshot
    var fields: [String: String]
    var heightCM: Double?
    var pending: PendingPreferences?
}
enum PreferencesError: LocalizedError {
    case message(String)
    var errorDescription: String? {
        if case .message(let message) = self { return message }
        return nil
    }
}

protocol PreferencesDraftStorage {
    func read(_ key: String) throws -> Data?
    func write(_ data: Data, key: String) throws
}
struct DefaultsPreferencesStorage: PreferencesDraftStorage {
    let defaults: UserDefaults
    func read(_ key: String) throws -> Data? {
        guard let value = defaults.object(forKey: key) else { return nil }
        guard let data = value as? Data else { throw PreferencesError.message("Unreadable saved preferences") }
        return data
    }
    func write(_ data: Data, key: String) throws {
        defaults.set(data, forKey: key)
        guard defaults.data(forKey: key) == data else { throw PreferencesError.message("Preferences could not be saved on this device.") }
    }
}

@MainActor
final class PreferencesStore: ObservableObject {
    @Published private(set) var draft: PreferencesDraft?
    @Published private(set) var conflict: PreferencesSnapshot?
    @Published private(set) var accepted: PreferencesSnapshot?
    @Published private(set) var isLoading = false
    @Published private(set) var isSaving = false
    @Published private(set) var isReadable = true
    @Published private(set) var error: String?
    @Published private(set) var message: String?
    let accountID: String
    let storageKey: String
    private let api: APIClient
    private let storage: any PreferencesDraftStorage
    private let ownerIsActive: () -> Bool
    private let onAccepted: (PreferencesSnapshot) -> Void
    private var isActive = true
    private var isOwned: Bool { isActive && ownerIsActive() }

    func activate() { isActive = true }
    func stop() { isActive = false }

    var isLocked: Bool { isSaving || draft?.pending != nil || !isReadable }
    var hasEdits: Bool { draft.map { $0.fields != PreferenceFields.fields($0.base) } ?? false }

    init(accountID: String, api: APIClient = .shared, defaults: UserDefaults = .standard,
         storage: (any PreferencesDraftStorage)? = nil,
         ownerIsActive: @escaping () -> Bool,
         onAccepted: @escaping (PreferencesSnapshot) -> Void = { _ in }) {
        self.accountID = accountID
        self.api = api
        self.storage = storage ?? DefaultsPreferencesStorage(defaults: defaults)
        self.ownerIsActive = ownerIsActive
        self.onAccepted = onAccepted
        storageKey = "preferences.draft.v1.\(api.storageNamespace).\(accountID)"
        do {
            if let data = try self.storage.read(storageKey) {
                let saved = try JSONDecoder().decode(PreferencesDraft.self, from: data)
                try validate(saved.base)
                guard saved.version == 1, saved.accountID == accountID,
                      PreferenceFields.definitions.allSatisfy({ saved.fields[$0.key] != nil }),
                      saved.heightCM == nil || saved.heightCM!.isFinite,
                      saved.pending == nil || (!saved.pending!.id.isEmpty && saved.pending!.body.baseRevision >= 0 && !saved.pending!.body.changes.isEmpty)
                else { throw PreferencesError.message("Unreadable saved preferences") }
                draft = saved
                message = "Your draft is saved on this iPhone. Review it before saving."
            }
        } catch {
            isReadable = false
            self.error = "Saved preferences could not be opened. Your draft has been kept. Keep the app installed and try again."
        }
    }
    private func validate(_ snapshot: PreferencesSnapshot) throws {
        guard snapshot.schemaVersion == 1, snapshot.accountID == accountID,
              snapshot.user.id == accountID, snapshot.revision >= 0 else {
            throw PreferencesError.message("The preferences response could not be read. Your draft is still here.")
        }
    }
    private func assertOwned() throws {
        try Task.checkCancellation()
        guard isOwned else { throw CancellationError() }
    }
    private func persist(_ next: PreferencesDraft) throws {
        try assertOwned()
        do { try storage.write(JSONEncoder().encode(next), key: storageKey) } catch {
            draft = next
            throw PreferencesError.message("This iPhone could not save your draft. Free some storage before sending it. Your open answers are kept.")
        }
        draft = next
    }
    private func fail(_ failure: Error) {
        guard isOwned, !(failure is CancellationError) else { return }
        error = failure.localizedDescription
    }
    private func adopt(_ snapshot: PreferencesSnapshot) throws {
        try assertOwned()
        try validate(snapshot)
        try persist(PreferencesDraft(version: 1, accountID: accountID, base: snapshot,
                                     fields: PreferenceFields.fields(snapshot), heightCM: snapshot.heightCM))
        accepted = snapshot
        conflict = nil
        error = nil
        message = "Preferences are up to date."
        onAccepted(snapshot)
    }
    func edit(_ key: String, value: String) {
        guard var next = draft, !isLocked, isOwned else { return }
        do {
            if key == "height" {
                if value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { next.heightCM = nil } else if let number = UserEnteredNumber.parse(value) {
                    next.heightCM = number * (next.fields["unitSystem"] == "imperial" ? 2.54 : 1)
                }
            }
            if key == "unitSystem" {
                let height = next.fields["height"] ?? ""
                guard height.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || UserEnteredNumber.parse(height) != nil else {
                    throw PreferencesError.message("Correct the height before changing display units.")
                }
                next.fields["height"] = PreferenceFields.heightText(next.heightCM, units: value)
            }
            next.fields[key] = value
            try persist(next)
            error = nil
            message = "Draft saved on this iPhone. Save preferences to apply it."
        } catch { fail(error) }
    }
    func load() async {
        guard !isLoading, !isSaving, isReadable, isOwned else { return }
        isLoading = true
        error = nil
        defer { isLoading = false }
        do { try await refresh() } catch { fail(error) }
    }
    private func fetch() async throws -> PreferencesSnapshot {
        let remote: PreferencesSnapshot = try await api.request("GET", path: "/api/preferences", expectedAccountID: accountID)
        try assertOwned()
        try validate(remote)
        return remote
    }
    private func refresh() async throws {
        if draft?.pending != nil { try await sendPending() }
        let remote = try await fetch()
        if let draft, hasEdits, remote.revision != draft.base.revision {
            conflict = remote
            error = nil
            message = "Preferences changed on another device. Review the differences."
        } else if !hasEdits { try adopt(remote) } else {
            accepted = remote
            onAccepted(remote)
            message = "Your draft is saved on this iPhone. Review it before saving."
        }
    }
    private func sendPending() async throws {
        guard var current = draft, let pending = current.pending else { return }
        try persist(current)
        do {
            let result: PreferencesSnapshot = try await api.request("PATCH", path: "/api/preferences", body: pending.body,
                                                                   operationID: pending.id, expectedAccountID: accountID)
            try adopt(result)
        } catch APIError.serverError(let status, let message) where status == 400 || status == 409 {
            current.pending = nil
            try persist(current)
            if status == 409 {
                conflict = try await fetch()
                self.message = "Preferences changed on another device. Review the differences."
                return
            }
            throw APIError.serverError(status, message)
        }
    }
    func save() async {
        guard !isLoading, !isSaving, isReadable, conflict == nil, isOwned, var current = draft else { return }
        isSaving = true
        error = nil
        defer { isSaving = false }
        do {
            if current.pending == nil {
                let changes = try PreferenceFields.changes(current)
                guard !changes.isEmpty else { message = "No unsaved changes."; return }
                current.pending = PendingPreferences(id: UUID().uuidString,
                                                     body: PreferenceMutation(baseRevision: current.base.revision, changes: changes))
                try persist(current)
            }
            try await sendPending()
            if conflict == nil {
                // An idempotent replay may acknowledge an older revision.
                // Fetch the current account before announcing completion.
                try await refresh()
                if conflict == nil { message = "Preferences saved." }
            }
        } catch { fail(error) }
    }
    func resolve(useServer: Bool) async {
        guard let remote = conflict, var current = draft, !isSaving, !isLoading, isOwned else { return }
        do {
            if useServer { try adopt(remote); return }
            let changes = try PreferenceFields.changes(current)
            var merged = remote
            for (key, value) in changes {
                if case .object(let patch) = value {
                    var group: [String: JSONValue] = [:]
                    if case .object(let existing) = merged.values[key] { group = existing }
                    group.merge(patch) { _, new in new }
                    merged.values[key] = .object(group)
                } else { merged.values[key] = value }
            }
            current.base = remote
            current.fields = PreferenceFields.fields(merged)
            current.heightCM = merged.heightCM
            current.pending = nil
            try persist(current)
            conflict = nil
            error = nil
            await save()
        } catch { fail(error) }
    }
}

File: apps/web/src/components/Profile/preferencesStore.ts
import API_CONFIG from '../../config';
import { api, ApiError, getSessionScope } from '../../lib/api';
import { operationID } from '../../lib/ids';
import {
  changesFor,
  fieldsFor,
  heightText,
  number,
  signature,
  labels,
  type Preferences,
  type Value,
} from './preferences';

interface Pending {
  id: string;
  body: { base_revision: number; changes: Record<string, Value> };
}
interface Draft {
  version: 1;
  accountID: string;
  base: Preferences;
  fields: Record<string, string>;
  heightCM: number | null;
  pending: Pending | null;
  updatedAt: number;
}
interface State {
  draft: Draft | null;
  loading: boolean;
  saving: boolean;
  readable: boolean;
  error: string | null;
  message: string | null;
  conflict: Preferences | null;
  accepted: Preferences | null;
}
let branch: { current: string; previous: string | null } | undefined;

export class PreferencesStore {
  private state: State = {
    draft: null,
    loading: false,
    saving: false,
    readable: true,
    error: null,
    message: null,
    conflict: null,
    accepted: null,
  };
  private listeners = new Set<() => void>();
  private prefix: string;
  private key: string;
  private stopped = false;
  private controller = new AbortController();
  private task: Promise<void> | null = null;
  constructor(
    readonly accountID: string,
    private scope: string
  ) {
    this.prefix = `exerly.preferences.v1:${API_CONFIG.BASE_URL}:${accountID}:`;
    if (!branch) {
      branch = { current: operationID(), previous: null };
      try {
        branch.previous = sessionStorage.getItem('exerly.preferences.branch');
        sessionStorage.setItem('exerly.preferences.branch', branch.current);
      } catch {
        /* Each document still has its own storage record. */
      }
    }
    this.key = this.prefix + branch.current;
    try {
      const drafts = this.savedDrafts();
      const previous =
        drafts.find(({ key }) => key === this.key) ??
        drafts.find(({ key }) => key === this.prefix + branch!.previous) ??
        drafts[0];
      if (previous) {
        this.state.draft = previous.draft;
        localStorage.setItem(this.key, JSON.stringify(previous.draft));
      }
    } catch {
      this.state.readable = false;
      this.state.error =
        'Saved preferences could not be opened. Existing drafts have been kept. Check site storage and reload.';
    }
  }
  subscribe = (listener: () => void) => {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  };
  getSnapshot = () => this.state;
  private publish(patch: Partial<State>) {
    if (this.stopped) return;
    this.state = { ...this.state, ...patch };
    for (const listener of this.listeners) listener();
  }
  private owned() {
    return !this.stopped && getSessionScope() === this.scope;
  }
  private assertOwned() {
    if (!this.owned()) throw new DOMException('Account changed', 'AbortError');
  }
  private validate(snapshot: Preferences) {
    if (
      !snapshot ||
      snapshot.schema_version !== 1 ||
      snapshot.account_id !== this.accountID ||
      !Number.isInteger(snapshot.revision) ||
      snapshot.revision < 0 ||
      !snapshot.values ||
      Array.isArray(snapshot.values) ||
      snapshot.user?._id !== this.accountID
    )
      throw new Error(
        'The saved preferences response could not be read. Your draft is still here.'
      );
  }
  private persist(draft: Draft) {
    this.assertOwned();
    try {
      localStorage.setItem(this.key, JSON.stringify(draft));
    } catch {
      this.publish({ draft });
      throw new Error(
        'This browser could not save your draft. Free some storage before sending it. Your open answers are kept.'
      );
    }
    this.publish({ draft });
  }
  private fail(error: unknown) {
    if (!this.owned() || (error instanceof Error && error.name === 'AbortError')) return;
    this.publish({
      error:
        error instanceof Error
          ? error.message
          : 'Could not load preferences. Retry when connected.',
    });
  }
  private async request<T>(send: (signal: AbortSignal) => Promise<T>): Promise<T> {
    const source = this.controller.signal;
    const controller = new AbortController();
    let timedOut = false;
    const cancel = () => controller.abort();
    if (source.aborted) cancel();
    source.addEventListener('abort', cancel, { once: true });
    const timer = window.setTimeout(() => {
      timedOut = true;
      cancel();
    }, 15000);
    try {
      return await send(controller.signal);
    } catch (error) {
      if (timedOut)
        throw new Error(
          'The connection took too long. Your draft is saved. Retry to check the result.'
        );
      throw error;
    } finally {
      window.clearTimeout(timer);
      source.removeEventListener('abort', cancel);
    }
  }
  private adopt(snapshot: Preferences) {
    this.validate(snapshot);
    this.persist({
      version: 1,
      accountID: this.accountID,
      base: snapshot,
      fields: fieldsFor(snapshot),
      heightCM: typeof snapshot.values.height === 'number' ? snapshot.values.height : null,
      pending: null,
      updatedAt: Date.now(),
    });
    this.publish({
      accepted: snapshot,
      conflict: null,
      message: 'Preferences are up to date.',
      error: null,
    });
  }
  start() {
    this.stopped = false;
    this.controller = new AbortController();
    void (this.task ?? Promise.resolve()).finally(() => this.load());
  }
  stop() {
    this.stopped = true;
    this.controller.abort();
  }
  private savedDrafts(): { key: string; draft: Draft }[] {
    return Object.keys(localStorage)
      .filter((key) => key.startsWith(this.prefix))
      .map((key) => {
        const draft = JSON.parse(localStorage.getItem(key)!) as Draft;
        if (
          draft.version !== 1 ||
          draft.accountID !== this.accountID ||
          !draft.fields ||
          Object.keys(labels).some((key) => typeof draft.fields[key] !== 'string') ||
          (draft.heightCM !== null && !Number.isFinite(draft.heightCM)) ||
          !Number.isFinite(draft.updatedAt) ||
          Object.values(draft.fields).some((value) => typeof value !== 'string') ||
          (draft.pending &&
            (typeof draft.pending.id !== 'string' ||
              !draft.pending.id ||
              !Number.isInteger(draft.pending.body?.base_revision) ||
              draft.pending.body.base_revision < 0 ||
              !draft.pending.body?.changes ||
              Array.isArray(draft.pending.body.changes) ||
              !Object.keys(draft.pending.body.changes).length))
        )
          throw new Error('Unreadable saved preferences');
        this.validate(draft.base);
        return { key, draft };
      })
      .sort((a, b) => b.draft.updatedAt - a.draft.updatedAt);
  }
  otherDrafts() {
    try {
      const seen = new Set([signature(this.state.draft?.fields)]);
      return this.savedDrafts().filter(({ key, draft }) => {
        const value = signature(draft.fields);
        if (key === this.key || seen.has(value)) return false;
        seen.add(value);
        return true;
      });
    } catch {
      return [];
    }
  }
  resume(key: string) {
    const draft = this.state.draft;
    if (!draft || draft.pending || this.state.saving || this.state.loading) return;
    try {
      const found = this.savedDrafts().find((entry) => entry.key === key);
      if (!found) return;
      // A recovered, unconfirmed save must retain the original revision and
      // operation. Copying its fields onto today's revision could overwrite a
      // later device edit without ever resolving that save.
      this.persist({
        ...found.draft,
        updatedAt: Date.now(),
      });
      this.publish({
        conflict: null,
        error: null,
        message: 'Recovered draft. Review it before saving.',
      });
      void this.load();
    } catch (error) {
      this.fail(error);
    }
  }
  edit(key: string, value: string) {
    const draft = this.state.draft;
    if (!draft || !this.state.readable || draft.pending || this.state.saving) return;
    try {
      const next = { ...draft, fields: { ...draft.fields, [key]: value }, updatedAt: Date.now() };
      if (key === 'height') {
        const parsed = number(value);
        if (parsed != null || !value.trim())
          next.heightCM =
            parsed == null ? null : parsed * (draft.fields.unitSystem === 'imperial' ? 2.54 : 1);
      }
      if (key === 'unitSystem') {
        if (draft.fields.height.trim() && number(draft.fields.height) == null)
          throw new Error('Correct the height before changing display units.');
        next.fields.height = heightText(next.heightCM, value);
      }
      this.persist(next);
      this.publish({
        error: null,
        message: 'Draft saved in this browser. Save preferences to apply it.',
      });
    } catch (error) {
      this.fail(error);
    }
  }
  load(): Promise<void> {
    if (this.task) return this.task;
    if (!this.owned() || !this.state.readable || this.state.saving) return Promise.resolve();
    this.publish({ loading: true, error: null });
    this.task = this.refresh()
      .catch((error) => this.fail(error))
      .finally(() => {
        this.task = null;
        this.publish({ loading: false });
      });
    return this.task;
  }
  private async refresh() {
    if (this.state.draft?.pending) await this.sendPending();
    const remote = await this.request((signal) =>
      api.get<Preferences>('/api/preferences', { signal })
    );
    this.assertOwned();
    this.validate(remote);
    const draft = this.state.draft;
    if (
      draft &&
      signature(draft.fields) !== signature(fieldsFor(draft.base)) &&
      remote.revision !== draft.base.revision
    ) {
      this.publish({
        conflict: remote,
        error: null,
        message: 'Preferences changed on another device. Review the differences.',
      });
    } else if (!draft || signature(draft.fields) === signature(fieldsFor(draft.base)))
      this.adopt(remote);
    else
      this.publish({
        accepted: remote,
        message: 'Your draft is saved locally. Review it before saving.',
      });
  }
  private async sendPending() {
    const draft = this.state.draft;
    if (!draft?.pending) return;
    this.persist(draft);
    try {
      const pending = draft.pending;
      const response = await this.request((signal) =>
        api.patch<Preferences>('/api/preferences', pending.body, {
          signal,
          operationID: pending.id,
        })
      );
      this.assertOwned();
      this.adopt(response);
    } catch (error) {
      if (error instanceof ApiError && [400, 409].includes(error.status)) {
        this.persist({ ...draft, pending: null });
        if (error.status === 409) {
          const remote = await this.request((signal) =>
            api.get<Preferences>('/api/preferences', { signal })
          );
          this.assertOwned();
          this.validate(remote);
          this.publish({ conflict: remote, error: null });
          return;
        }
      }
      throw error;
    }
  }
  async save() {
    if (
      !this.owned() ||
      !this.state.readable ||
      this.state.saving ||
      this.state.conflict ||
      !this.state.draft
    )
      return;
    if (this.task) await this.task;
    // A refresh can discover a conflict or outlive the signed-in account.
    // Two save clicks can also have been waiting on the same refresh.
    if (
      !this.owned() ||
      !this.state.readable ||
      this.state.saving ||
      this.state.conflict ||
      !this.state.draft
    )
      return;
    this.publish({ saving: true, error: null });
    try {
      const draft = this.state.draft!;
      if (!draft.pending) {
        const changes = changesFor(draft.base, draft.fields, draft.heightCM);
        if (!Object.keys(changes).length) {
          this.publish({ message: 'No unsaved changes.' });
          return;
        }
        this.persist({
          ...draft,
          pending: { id: operationID(), body: { base_revision: draft.base.revision, changes } },
        });
      }
      await this.sendPending();
      if (!this.state.conflict) {
        await this.refresh();
        if (!this.state.conflict) this.publish({ message: 'Preferences saved.' });
      }
    } catch (error) {
      this.fail(error);
    } finally {
      this.publish({ saving: false });
    }
  }
  async resolve(useServer: boolean) {
    const remote = this.state.conflict;
    const draft = this.state.draft;
    if (!remote || !draft || !this.owned() || this.state.saving || this.state.loading) return;
    try {
      if (useServer) this.adopt(remote);
      else {
        const changes = changesFor(draft.base, draft.fields, draft.heightCM);
        const values = structuredClone(remote.values);
        for (const [key, value] of Object.entries(changes))
          values[key] =
            value && typeof value === 'object' && !Array.isArray(value)
              ? { ...(values[key] as Record<string, Value>), ...value }
              : value;
        const merged = { ...remote, values };
        this.persist({
          ...draft,
          base: remote,
          fields: fieldsFor(merged),
          heightCM: typeof values.height === 'number' ? values.height : null,
          pending: null,
        });
        this.publish({ conflict: null, error: null });
        await this.save();
      }
    } catch (error) {
      this.fail(error);
    }
  }
}
